### PR TITLE
Add spi service initial setup and updated few pair screen ui tweaks

### PR DESCRIPTION
**Pull request by mx51damon**
* Add SPI service initial setup (with unit tests)
* Update a few pair screen ui tweaks

### CODE REVIEW

**Review by mattgresham — APPROVED** [review]
Nice work, the library is a complex beast!

I've left a couple of comments for a few small things to think about.

File: src/definitions/constants/commonConfigs.ts [hunk 1]
```diff
@@ -10,8 +10,41 @@ export const TEXT_FORM_VALIDATION_POS_ID_TEXTFIELD = 'Please enter a POS ID.';
 export const TEXT_FORM_VALIDATION_PROVIDER_TEXTFIELD = 'Please enter a supported provider.';
 export const TEXT_FORM_VALIDATION_SERIAL_NUMBER_TEXTFIELD = 'Please enter a 9 digits Serial number.';
 // pair status text
-export const TEXT_STATUS_UNPAIRED = 'Idle';
-export const TEXT_STATUS_PAIRING = 'Pairing';
-export const TEXT_STATUS_PAIRED = 'Transaction';
-// Pair Flow drawer width 4/12 (based on design required)
+export const SPI_PAIR_FLOW = {
```
[the remark this answers is not in the document]
Comment (mx51damon) [reply src/definitions/constants/commonConfigs.ts:13]: Good to know Matt, I will update this part ~~

File: src/services/spiService.ts [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+    // spiDeviceAddressChanged
+    terminal.addEventListener(spiEvents.spiPairingFlowStateChanged, (event: Any) => {
+      this.print.log(`%cspiDeviceAddressChanged: ${JSON.stringify(event, null, 2)}`, `color: ${PRIMARY_THEME_COLOR}`);
+      // @Trigger redux action to update state for UI
+    });
+
+    // spiStatusChanged
+    terminal.addEventListener(spiEvents.spiStatusChanged, (event: Any) => {
+      if (event && event.detail && event.detail === SPI_PAIR_STATUS.CONNECTED) {
```
[the remark this answers is not in the document]
Comment (mx51damon) [reply src/services/spiService.ts:128]: Yeah, good catch, I will update this part ~~

File: src/services/spiService.ts [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+      this.print.log(`%cspiStatusChanged: ${CurrentStatus}`, `color: ${PRIMARY_THEME_COLOR}`);
+    });
+
+    // spiPairingFlowStateChanged event
+    terminal.addEventListener(spiEvents.spiPairingFlowStateChanged, (event: Any) => {
+      const { detail } = event;
+
+      if (detail?.ConfirmationCode && !detail.AwaitingCheckFromEftpos && detail.AwaitingCheckFromPos) {
+        terminal.spiClient.PairingConfirmCode();
+      }
```
[the remark this answers is not in the document]
Comment (mx51damon) [reply src/services/spiService.ts:143]: Good to know Matt, I will remove this code block ~~

File: src/services/spiService.ts [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+  readCurrentFlow(): Any {
+    // @Trigger redux action to update state for UI
+    return this.readCurrentInstance().currentTxFlowStateOverride || this.readCurrentInstance().CurrentTxFlowState;
+  }
+
+  // Pair related functionalities
+  spiPairTerminal(dispatch: AppDispatch): void {
+    this.readCurrentInstance().spiClient.Pair();
+
+    const currentUpdates = {
```
[the remark this answers is not in the document]
Comment (mx51damon) [reply src/services/spiService.ts:217]: Yup, looks like this one will return an object, cool, I will use this object then ~~

File: src/services/spiService.ts [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+      return terminal;
+    } catch (error) {
+      const errorOutput = error || 'Error occurred during terminal configuration setup';
+
+      this.print.error(`%c ${errorOutput}`, `color: ${PRIMARY_ERROR_COLOR};`);
+      throw errorOutput;
+    }
+  }
+
+  createOrUpdateTerminal(instanceId: string, pairFormSettings: IPairFormSettings): ITerminalSetups {
```
[the remark this answers is not in the document]
Comment (mx51damon) [reply src/services/spiService.ts:107]: Thanks Matt, I will rename the function and may I confirm is that only need `createTerminal()` when we do **pair** session? when we go to terminal details page, we call `updateTerminal()`, is that my understanding correct?

I was thinking shall we take all listeners out as a function and call that function inside `createOrUpdateTerminal`, then we rename it as `createTerminal` and also we generate another function called `updateTerminal` without those listeners. What do you think?

[the remark this answers is not in the document]
Comment (mattgresham) [reply src/services/spiService.ts:107]: That sounds like a good idea. But before you do a whole bunch of refactoring for this, just keep it in mind for now and once you have the pairing screen and terminals list wired up you can circle back to this and see what you think makes sense

[the remark this answers is not in the document]
Comment (mx51damon) [reply src/services/spiService.ts:107]: Thanks Matt, I got the point, make it work first then refactor once we got the skeleton or frame ~~

**Pull request merged**